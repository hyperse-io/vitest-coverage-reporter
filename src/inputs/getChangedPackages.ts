import { Package } from '@manypkg/get-packages';
import { FileCoverageMode } from './getCoverageModeFrom';
import { getPullChanges } from './getPullChanges';
import { getWorkspacePackages } from './getWorkspacePackages';

export async function getChangedPackages(repoCwd: string) {
  const workspacePackages = await getWorkspacePackages(repoCwd);
  const changedPackages = new Set<Package>();
  const allChangedFiles = await getPullChanges(FileCoverageMode.All);

  for (const [dir, { name, relativeDir, version }] of workspacePackages) {
    const packageChanged = allChangedFiles.find(
      (s) => !!~s.indexOf(relativeDir)
    );

    if (packageChanged) {
      changedPackages.add({
        dir,
        relativeDir,
        packageJson: {
          name,
          version,
        },
      });
    }
  }

  return [...changedPackages];
}
