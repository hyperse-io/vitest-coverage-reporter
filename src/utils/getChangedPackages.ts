import { getPackages, Package } from '@manypkg/get-packages';

export async function getChangedPackages(
  cwd: string,
  workspacePackages: Map<string, { name: string; version: string }>
) {
  const { packages } = await getPackages(cwd);
  const changedPackages = new Set<Package>();

  for (const pkg of packages) {
    const previousPackages = workspacePackages.get(pkg.dir);
    if (previousPackages?.version !== pkg.packageJson.version) {
      changedPackages.add(pkg);
    }
  }

  return [...changedPackages];
}
