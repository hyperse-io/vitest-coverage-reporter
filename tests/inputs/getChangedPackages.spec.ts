import { join } from 'node:path';
import { getChangedPackages } from '../../src/inputs/getChangedPackages.js';
import * as pullChanges from '../../src/inputs/getPullChanges.js';

describe('getChangedPackages()', () => {
  const fixtureCwd = join(__dirname, 'fixtures');
  beforeEach(() => {
    vi.spyOn(pullChanges, 'getPullChanges').mockResolvedValue([
      'packages/package1/src/index.ts',
      'packages/package2/src/index.ts',
      'website/index.ts',
    ]);
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it('Should correct resolve file changed packages', async () => {
    const changedPackages = await getChangedPackages(fixtureCwd);
    for (const { dir: _, ...restProps } of changedPackages) {
      expect(restProps).toBeDefined();

      expect(restProps).toMatchSnapshot();
    }
  });
});
