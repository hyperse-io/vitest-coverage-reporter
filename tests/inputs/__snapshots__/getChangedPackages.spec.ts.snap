// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getChangedPackages() > Should correct resolve file changed packages 1`] = `
{
  "packageJson": {
    "name": "fixture-package1",
    "version": "1.0.0",
  },
  "relativeDir": "packages/package1",
}
`;

exports[`getChangedPackages() > Should correct resolve file changed packages 2`] = `
{
  "packageJson": {
    "name": "fixture-package2",
    "version": "1.0.0",
  },
  "relativeDir": "packages/package2",
}
`;

exports[`getChangedPackages() > Should correct resolve file changed packages 3`] = `
{
  "packageJson": {
    "name": "fixture-website",
    "version": "1.0.0",
  },
  "relativeDir": "website",
}
`;
